### ISSUE TITLE
Implement two-way replay protection

### ISSUE THREAD

**Issue by prusnak** [BTCGPU/BTCGPU#18]
This is necessary in order not to cause havoc. 

Bitcoin Cash implemented this by using SIGHASH_FORKID and I recommend you doing the same, as it would be trivial to implement your coin into other wallets that already support Bitcoin Cash.

More info:

* https://github.com/Bitcoin-ABC/bitcoin-abc/issues/24
* https://github.com/Bitcoin-UAHF/spec/pull/17

**Comment by leto** [comment]
Is the current plan to launch the mainnet on Nov 1st, even if this is not implemented? That seems dangerous.

**Comment by h4x3rotab** [comment]
Work in progress. Reply protection is essential for the launch. In case there's anything wrong with it, the launch date could be postponed.

**Comment by leto** [comment]
@h4x3rotab awesome, thanks for clarifying that! People that work on other Equihash coins, like me, are rooting for y'all! Best of luck.

**Comment by bithernet** [comment]
@h4x3rotab , have you implemented two-way replay protection? Did you implement it by using SIGHASH_FORKID like Bitcoin Cash?

Any docs?

**Comment by leto** [comment]
:+1: to more docs about the exact method of providing 2-way replay protection

**Comment by drewdotpro** [comment]
Less than a week away with no replay protection code pushed? Developers for wallets like Trezor will want to support Bitcoin Gold but can't start development until you show your replay protection implementation.

**Comment by ronaldobini** [comment]
We here in Brazil also only waiting the push of this code

**Comment by ronaldobini** [comment]
we here in Brazil just waiting it too.

**Comment by Paulnternet** [comment]
It's might play out as a full on attack on the legacy chain. Choose a side and buckle up.

**Comment by prusnak** [comment]
Also remember it takes days/weeks to release a new firmware for hardware wallets with support for new coin, so don't keep major technical changes for later.

**Comment by usefree** [comment]
So what? Still not solved? How YoBit is going to list it?

**Comment by h4x3rotab** [comment]
In progress

**Comment by prusnak** [comment]
Link to public branch where this is being developed so we can comment?

**Comment by leto** [comment]
This seems to be under development in private. Somewhat concerning.

**Comment by prusnak** [comment]
Or not being developed at all ...

**Comment by usefree** [comment]
@prusnak please, if i understand correctly, after this issue will be solved it will become possible to implement Trezor web-wallet support?

**Comment by prusnak** [comment]
This and address change in https://github.com/BTCGPU/BTCGPU/issues/17

**Comment by saleemrashid** [comment]
Will you use `SIGHASH_FORKID`? If not, why will you not? Most wallets (including web wallets and hardware wallets) support `SIGHASH_FORKID` so it would be incredibly reckless to choose another replay protection scheme if you want support for Bitcoin Gold implemented and released quickly.

**Comment by h4x3rotab** [comment]
> Will you use SIGHASH_FORKID? If not, why will you not? Most wallets (including web wallets and hardware wallets) support SIGHASH_FORKID so it would be incredibly reckless to choose another replay protection scheme if you want support for Bitcoin Gold implemented and released quickly.

Yes, the same 2-way replay protection as Bitcoin Cash. But we will not implement their legacy OP_RETURN protection.

**Comment by saleemrashid** [comment]
> Yes, the same 2-way replay protection as Bitcoin Cash. But we will not implement their legacy OP_RETURN protection.

That _sounds_ excellent, but where is the code? To use your own words, this has been "work in progress" for weeks.

**Comment by h4x3rotab** [comment]
Too many features requested. We have to do it one by one. The original plan is to launch a small scale pre-testnet without replay protection, add it later, and then launch another testnet and the mainnet. Now we are putting all the bandwidth on this.

**Comment by saleemrashid** [comment]
@h4x3rotab When exactly is the testnet going to be launched then? If you want support released quickly, it would be a good idea to have a testnet running at least a few days before the mainnet yet the fork is supposed to happen in about 550 blocks.

**Comment by h4x3rotab** [comment]
It's the snapshot block height. The full launch is targeted on Nov 1 and could be postponed to ensure the safety if necessary.

**Comment by saleemrashid** [comment]
> It's the snapshot block height.

Why not just fork when the launch happens?

**Comment by StarbuckBG** [comment]
There is a bounty for this: #51

**Comment by h4x3rotab** [comment]
> Why not just fork when the launch happens?

Tight timeline, and there will always be a gap due to the PoW change. A snapshot is the best choice to ensure a smooth launch. Actually we've repeated this for several times but you might missed it.

**Comment by prusnak** [comment]
Why do you just started a bounty if this has been a work in progress since at least 16 days ago? This does not look good at all ...

**Comment by joncursi** [comment]
T-minus 3.5 days according to the countdown on http://btcgpu.org/. Oh boy...

**Comment by davidmann4** [comment]
Will there be a snapshot in 3.5 days or will the fork just happen when it has replay protection?

**Comment by davidmann4** [comment]
Nevermind it got answered serval times: you do a snapshot. Cool.

**Comment by ileathan** [comment]
Wait so no one has implemented this? What even is the bounty? I dont even know if I can do it in this time anymore though..

**Comment by ileathan** [comment]
Well theres like 7 hours left now? GL lol... >_<

**Comment by prusnak** [comment]
![1y4h92](https://user-images.githubusercontent.com/42201/31911968-b243a4a4-b842-11e7-9424-c3244ddab3ab.jpg)

**Comment by lichtamberg** [comment]
@h4x3rotab Why dont you postpone the fork if all these features are not implemented yet? It looks not very serious to start something, which is not properly tested and reviewed (nor even implemented)?

**Comment by ronaldobini** [comment]
Ow man.. Just think! This is a fork not the launch of the coin.. They will only take a snapshot just it. The deposit and withdrawal will only be available a bunch of days later

**Comment by lichtamberg** [comment]
So why now? and not at the launch of the coin?

**Comment by Mediazone** [comment]
Tomorrow at 12 UTC they will take the BTC blockchain as a starting point for Bitcoin Gold. Then they will premise on the new chain with the new not released wallet (without replay protection and bugs). After that a public wallet will be available. It could take some time IMHO

**Comment by ronaldobini** [comment]
Is simple. They chose this block to fork. They didn't finished the replay protection on time, but is all ready for the fork then they have time to implement, and is a lot easier than other things like pow that they have implemented. So they already stated that they won't launch without the protection. But now there is other priorities. So please stop complaining like noobs. They will implement that same protection than BCash almost crtl c crtl V and boom, they can launch without problems.. Now please relax guys... And will be just another bitcoin free

**Comment by unknownids** [comment]
yes, the people complaining here are the noobs.

**Comment by mryellow** [comment]
>  They will implement that same protection than BCash almost crtl c crtl V and boom, they can launch without problems.. 

Actually some last minute dodgy negative version number hack.

https://github.com/BTCGPU/BTCGPU/pull/62

**Comment by Ivans1310** [comment]
replay protection already implemented ?, and if so, when trezor would enable a wallet for bitcoin Gold?

**Comment by dudzcom** [comment]
Is this done?

**Comment by Ivans1310** [comment]
I think not, because they still do not deliver them supposedly until November

**Comment by briannyeko** [comment]
Replay protection done?

**Comment by StarbuckBG** [comment]
@briannyeko #83 
Will close that issue. Please refer to this issue: #51

**Comment by lacostenycoder** [comment]
IMHO this is possible the worst rollout of a hard fork imaginable. Anyone who held their BTC on a Ledger Nano S or Trezor like myself basically lost any chance at selling off their BTG when it still had any value.  Hope you core dev(s) really only 1 main dev (how is this possible?) made out like a bandit and sold your pre-mined coins before it tanked.  I don't mean to sound negative and I hope there comes some real value to this eventually. In the meantime... 

![screen shot 2017-10-26 at 2 48 36 pm](https://user-images.githubusercontent.com/6077905/32071439-51c934b4-ba5d-11e7-8008-9942767737bd.jpg)

**Comment by hashmonkey** [comment]
100% agree with @lacostenycoder 

this is a fucking joke.  vapourware it should be called from now on...

![22852178_10208152641972438_2174978924627953445_n](https://user-images.githubusercontent.com/28983492/32071841-7f994cfa-ba88-11e7-93d3-5a0c6e020d85.jpg)

**Comment by briannyeko** [comment]
hahaha you guys are hillarious!!

**Comment by StarbuckBG** [comment]
Stop spamming FFS. When you want to spam or hate, create your repo and do it there. Some people here are trying to get the work done as you have no other job than just spamming.

We want to have clear emails from github in order to be aware if there are real issues and that way you are making it really hard…
Please respect my work as I don’t spam your repos and making your work hard!

**Comment by rikur** [comment]
Feel free to spam my repos whenever I let out blatant lies!
